airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/deser/AirbyteMessageDeserializer.kt

```
@@ -3,37 +3,65 @@
      * * entire serialized message string when message is a valid State Message
      * * serialized AirbyteRecordMessage when message is a valid Record Message
      *
      * @param message the string to deserialize
-     * @return PartialAirbyteMessage if the message is valid, empty otherwise
+     * @return PartialAirbyteMessage if the message is valid, null if there was a recoverable error
      */
     fun deserializeAirbyteMessage(
         message: String?,
-    ): PartialAirbyteMessage {
+    ): PartialAirbyteMessage? {
         // TODO: This is doing some sketchy assumptions by deserializing either the whole or the
         // partial based on type.
         // Use JsonSubTypes and extend StdDeserializer to properly handle this.
         // Make immutability a first class citizen in the PartialAirbyteMessage class.
         val partial =
-            Jsons.tryDeserializeExact(message, PartialAirbyteMessage::class.java).orElseThrow {
-                RuntimeException("Unable to deserialize PartialAirbyteMessage.")
+            try {
+                try {
+                    Jsons.deserializeExactUnchecked(message, PartialAirbyteMessage::class.java)
+                } catch (e: ValueInstantiationException) {
+                    // This is a hack to catch unrecognized message types. Jackson supports
+                    // the equivalent via annotations, but we cannot use them because the
+                    // AirbyteMessage
+                    // is generated from json-schema.
+                    val pat =
+                        Regex(
+                            "Cannot construct instance of .*AirbyteMessage.Type., problem: ([_A-Z]+)"
+                        )
+                    val match = pat.find(e.message!!)
+                    if (match != null) {
+                        val unrecognized = match.groups[1]?.value
+                        logger.warn { "Unrecognized message type: $unrecognized" }
+                        return null
```

I'd rather throw a typed exception than return null here. This way, we also avoid all the !! everywhere we know this to be valid

---

Done.